modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/defragmentation/DefragIndexFactory.java

```diff
@@ -0,0 +1,262 @@
+            pageIoResolver(),
+            rootPage.pageId().pageId(),
+            rootPage.isAllocated(),
+            oldIdx.inlineSize(),
+            def.keyTypeSettings(),
+            null,
+            stats,
+            rowHndFactory,
+            null
+        );
+
+        final MetaPageInfo oldInfo = oldIdx.segment(segmentNum).metaInfo();
+
+        tree.copyMetaInfo(oldInfo);
+
+        tree.enableSequentialWriteMode();
+
+        return tree;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected RootPage rootPage(GridCacheContext<?, ?> ctx, String treeName, int segment) throws Exception {
+        return offheap.rootPageForIndex(ctx.cacheId(), treeName, segment);
+    }
+
+    /** */
+    private PageIoResolver pageIoResolver() {
+        return pageAddr -> {
+            PageIO io = PageIoResolver.DEFAULT_PAGE_IO_RESOLVER.resolve(pageAddr);
+
+            if (io instanceof BPlusMetaIO)
+                return io;
+
+            //noinspection unchecked,rawtypes,rawtypes
+            return wrap((BPlusIO)io, rowHndFactory.create(null, null));
+        };
+    }
+
+    /** */
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    static BPlusIO<IndexRow> wrap(BPlusIO<IndexRow> io, InlineIndexRowHandler rowHnd) {
+        assert io instanceof InlineIO;
+
+        if (io instanceof BPlusInnerIO) {
+            assert io instanceof AbstractInlineInnerIO
+                || io instanceof InlineInnerIO;
+
+            return new BPlusInnerIoDelegate((BPlusInnerIO<IndexRow>)io, rowHnd);
+        }
+        else {
+            assert io instanceof AbstractInlineLeafIO
+                || io instanceof InlineLeafIO;
+
+            return new BPlusLeafIoDelegate((BPlusLeafIO<IndexRow>)io, rowHnd);
+        }
+    }
+
+    /** */
+    private static <T extends BPlusIO<IndexRow> & InlineIO> IndexRow lookupRow(
+        InlineIndexRowHandler rowHnd,
+        long pageAddr,
+        int idx,
+        T io
+    ) {
+        long link = io.link(pageAddr, idx);
+
+        int off = io.offset(idx);
+
+        IndexKey[] keys = new IndexKey[rowHnd.indexKeyDefinitions().size()];
+
+        int fieldOff = 0;
+
+        for (int i = 0; i < rowHnd.inlineIndexKeyTypes().size(); i++) {
+            InlineIndexKeyType keyType = rowHnd.inlineIndexKeyTypes().get(i);
+
+            IndexKey key = keyType.get(pageAddr, off + fieldOff, io.inlineSize() - fieldOff);
+
+            fieldOff += keyType.inlineSize(key);
+
+            keys[i] = key;
+        }
+
+        return new IndexRowImpl(rowHnd, new CacheDataRowAdapter(link), keys);
+    }
+
+    /** */
+    private static class BPlusInnerIoDelegate<IO extends BPlusInnerIO<IndexRow> & InlineIO>
+        extends BPlusInnerIO<IndexRow> implements InlineIO {
+        /** */
+        private final IO io;
+
+        /** */
+        private final InlineIndexRowHandler rowHnd;
+
+        /** */
+        public BPlusInnerIoDelegate(IO io, InlineIndexRowHandler rowHnd) {
```

Comment:
public constructor for private inner class (the same below)

Reply:
fixed.

Reply:
fixed.